relayer/client.go

```diff
@@ -238,137 +234,182 @@ func CreateClient(ctx context.Context, src, dst *Chain, srcUpdateHeader, dstUpda
-		if err != nil {
+	var srcHeader, dstTrustedHeader provider.IBCHeader
+
+	eg, egCtx := errgroup.WithContext(ctx)
+	eg.Go(func() error {
+		return retry.Do(func() error {
+			var err error
+			srcHeader, err = src.ChainProvider.QueryIBCHeader(egCtx, srch)
 			return err
-		}
-		return nil
+		}, retry.Context(ctx), RtyAtt, RtyDel, RtyErr, retry.OnRetry(func(n uint, err error) {
+			src.log.Info(
+				"Failed to query IBC header when building update client message",
+				zap.String("client_id", dst.ClientID()),
+				zap.Uint("attempt", n+1),
+				zap.Uint("max_attempts", RtyAttNum),
+				zap.Error(err),
+			)
+		}))
+	})
+	eg.Go(func() error {
+		return retry.Do(func() error {
+			var err error
+			dstTrustedHeader, err = src.ChainProvider.QueryIBCHeader(egCtx, int64(dstClientState.GetLatestHeight().GetRevisionHeight())+1)
+			return err
+		}, retry.Context(ctx), RtyAtt, RtyDel, RtyErr, retry.OnRetry(func(n uint, err error) {
+			src.log.Info(
+				"Failed to query IBC header when building update client message",
+				zap.String("client_id", dst.ClientID()),
+				zap.Uint("attempt", n+1),
+				zap.Uint("max_attempts", RtyAttNum),
+				zap.Error(err),
+			)
+		}))
+	})
+
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	var updateHeader ibcexported.Header
+	if err := retry.Do(func() error {
+		var err error
+		updateHeader, err = src.ChainProvider.MsgUpdateClientHeader(srcHeader, dstClientState.GetLatestHeight().(clienttypes.Height), dstTrustedHeader)
+		return err
 	}, retry.Context(ctx), RtyAtt, RtyDel, RtyErr, retry.OnRetry(func(n uint, err error) {
-		c.log.Info(
-			"Failed to get IBC update headers",
+		src.log.Info(
+			"Failed to build update client header",
+			zap.String("client_id", dst.ClientID()),
 			zap.Uint("attempt", n+1),
 			zap.Uint("max_attempts", RtyAttNum),
 			zap.Error(err),
 		)
-		srch, dsth, _ = QueryLatestHeights(ctx, c, dst)
 	})); err != nil {
-		return err
+		return nil, err
 	}
 
-	srcUpdateMsg, err := c.ChainProvider.MsgUpdateClient(c.ClientID(), dstUpdateHeader)
+	// updates off-chain light client
+	return dst.ChainProvider.MsgUpdateClient(dst.ClientID(), updateHeader)
+}
+
+// UpdateClients updates clients for src on dst and dst on src given the configured paths
+func UpdateClients(
+	ctx context.Context,
+	src, dst *Chain,
+	memo string,
+) (err error) {
+	srch, dsth, err := QueryLatestHeights(ctx, src, dst)
 	if err != nil {
-		c.log.Debug(
-			"Failed to update source client",
-			zap.String("src_chain", c.ChainID()),
-			zap.Error(err),
-		)
 		return err
 	}
 
-	dstUpdateMsg, err := dst.ChainProvider.MsgUpdateClient(dst.ClientID(), srcUpdateHeader)
-	if err != nil {
-		dst.log.Debug(
-			"Failed to update destination client",
-			zap.String("dst_chain", dst.ChainID()),
-			zap.Error(err),
-		)
+	var srcMsgUpdateClient, dstMsgUpdateClient provider.RelayerMessage
+	eg, egCtx := errgroup.WithContext(ctx) // New errgroup because previous egCtx is canceled at this point.
```

Comment:
Comment is stale (there was no previous egCtx in scope).

Reply:
done